### PR TITLE
[Issue: 12] Deploy as -> Dashboard to Jupyter Dashboard Server

### LINKED ISSUE #12: Deploy as -> Dashboard to Jupyter Dashboard Server
Write a bundler that takes a notebook and POSTs it to a  `jupyter-incubator/dashboards_nodejs_app` instance (now called Jupyter Dashboard Server for simplicity). Reads the URL of the dashboard server from an environment variable set when the Jupyter Notebook Server is brought up (e.g., `DASHBOARD_SERVER_URL`).

Only posts the notebook for the time being, no additional dependencies and associated resources (because that's all the POST handler supports in the dashboard server).

* [x] Initially, do a PR that does the POST assuming there's no auth on the dashboard server.
* [x] Later, after jupyter-incubator/dashboards_nodejs_app#45 lands, do another PR that sends the proper auth token along in the POST. The auth token will be read out of the environment on the Notebook server side as well (`DASHBOARD_SERVER_TOKEN`).

### PR DESCRIPTION
**Pull request by nitind**
Basic notebook upload without associated resources

Fixes #12

### CODE REVIEW

File: dashboards_bundlers/server_upload/__init__.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    '''
+    # Get name of notebook from filename
+    notebook_basename = os.path.basename(abs_nb_path)
+    notebook_name = os.path.splitext(notebook_basename)[0]
+    if 'DASHBOARD_SERVER_URL' in os.environ:
+        server_url = url_path_join(os.getenv('DASHBOARD_SERVER_URL'), ENDPOINT, escape.url_escape(notebook_name, False))
+        with open(abs_nb_path, 'rb') as notebook:
+            headers = {}
+            if 'DASHBOARD_SERVER_TOKEN' in os.environ:
+                headers = headers + {'Authorization':'token ' + os.getenv('DASHBOARD_SERVER_TOKEN')}
```
Comment (parente) [inline dashboards_bundlers/server_upload/__init__.py]: `headers['Authorization'] = 'token ' + os.getenv('DASHBOARD_SERVER_TOKEN')`

File: dashboards_bundlers/server_upload/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+def bundle(handler, abs_nb_path):
+    '''
+    Uploads a notebook to a Jupyter Dashboard Server
+    '''
+    # Get name of notebook from filename
+    notebook_basename = os.path.basename(abs_nb_path)
+    notebook_name = os.path.splitext(notebook_basename)[0]
+    if 'DASHBOARD_SERVER_URL' in os.environ:
+        server_url = url_path_join(os.getenv('DASHBOARD_SERVER_URL'), ENDPOINT, escape.url_escape(notebook_name, False))
```
Comment (parente) [inline dashboards_bundlers/server_upload/__init__.py]: Not sure how stable our API will be. Drew's already changing it to clean it up. Might be we want the URL directly to the notebooks API endpoint. But for now, leave it.

File: dashboards_bundlers/server_upload/__init__.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        with open(abs_nb_path, 'rb') as notebook:
+            headers = {}
+            if 'DASHBOARD_SERVER_TOKEN' in os.environ:
+                headers = headers + {'Authorization':'token ' + os.getenv('DASHBOARD_SERVER_TOKEN')}
+            result = requests.post(server_url, files={'file': notebook}, headers=headers)
+            result.raise_for_status()
+    else:
+        access_log.debug('Can not deploy, DASHBOARD_SERVER_URL not set')
```
Comment (parente) [inline dashboards_bundlers/server_upload/__init__.py]: Bundler needs to finish the web response at the end. On success, response should be a redirect to the dashboard server URL with the `/notebooks/<name of notebook>` appended. (Again, this is going to change shortly in jupyter-incubator/dashboards_nodejs_app#48. When the DASHBOARD_SERVER_URL is not set, `handler.send_error` or raise a Tornado web error with status 500 and a custom message which the Jupyter error page will show in the page. When any other error occurs, let it propagate (like you're doing).

File: dashboards_bundlers/server_upload/__init__.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        with open(abs_nb_path, 'rb') as notebook:
+            headers = {}
+            if 'DASHBOARD_SERVER_TOKEN' in os.environ:
+                headers['Authorization'] = 'token ' + os.getenv('DASHBOARD_SERVER_TOKEN')
+            result = requests.post(dashboard_url, files={'file': notebook}, headers=headers)
+            result.raise_for_status()
+            handler.redirect(dashboard_url)
+    else:
+        access_log.debug('Can not deploy, DASHBOARD_SERVER_URL not set')
+        raise httpclient.HTTPError(500, message='No dashboard server configured')
```
Comment (parente) [inline dashboards_bundlers/server_upload/__init__.py]: I think you want http://tornado.readthedocs.org/en/latest/web.html#tornado.web.HTTPError, not httpclient error.

**Comment by nitind** [comment]
Updated with logging correction, DASHBOARD_SERVER_AUTH_TOKEN variable name to be more obviously related to the server's variable name, and the recently changed post and render URIs for the dashboard server.

File: dashboards_bundlers/extensionapp.py [hunk 5]
```diff
@@ -59,6 +59,11 @@ def start(self):
                     'label': 'PHP Dashboard bundle (.zip)',
                     'module_name': 'dashboards_bundlers.php_download',
                     'group': 'download'
+                },
+                'dashboards_server_upload': {
+                    'label': 'Dashboard on Dashboard Server',
```
Comment (parente) [inline dashboards_bundlers/extensionapp.py]: Maybe Dashboard on Jupyter Dashboard Server just to give a bit more of a hint of where it's going?

Comment (parente) [inline dashboards_bundlers/extensionapp.py]: Also it's a bit unfortunate that the menu item will always appear even if the backend is not configured properly. We might want to improve the bundler API over in jupyter_cms to account for this.

**Comment by parente** [comment]
I had trouble testing this. After setting up the env vars and running the dashboard server, when I click deploy, it hangs. It's probably my env I'm sure. I'll hold off on trying again until the Makefile gets updated to make testing easier.

File: dashboards_bundlers/extensionapp.py (hunk 5 above)
Comment (nitind) [inline dashboards_bundlers/extensionapp.py]: Changed.

**Comment by nitind** [comment]
Added the timeout and vars to the Makefile, then rechecked against the dashboards_nodejs_app demo-container.

**Comment by parente** [comment]
Tested manually. Works as advertised with and without auth enabled.

Noticed that 401 errors from the requests made to the dashboard server are always raised as 500 errors. Will open a follow on issue to fix that later.

**Pull request merged**